llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -914,3 +919,64 @@ bool SpecConstantsPass::collectSpecConstantDefaultValuesMetadata(
 
   return true;
 }
+
+void replaceSpecConstsWithDefaultValue(const Module &M,
+                                       const std::vector<CallInst *> &CIs) {
+  const DataLayout &DL = M.getDataLayout();
+  for (CallInst *CI : CIs) {
+    Value *Dst = CI->getArgOperand(0);
+    Value *Src = CI->getArgOperand(2);
+    Align DstAlign = DL.getABITypeAlign(Dst->getType());
+    Align SrcAlign = DL.getABITypeAlign(Src->getType());
+    Type *StructureType =
+        Src->stripPointerCasts()->getType()->getContainedType(0);
+    uint64_t SpecConstSize = DL.getTypeAllocSize(StructureType);
+
+    IRBuilder B(CI);
+    B.CreateMemCpy(Dst, DstAlign, Src, SrcAlign, SpecConstSize);
+    CI->removeFromParent();
+    CI->deleteValue();
+  }
+}
+
+static bool checkModuleContainsSpecConsts(const Module &M) {
+  for (const Function &F : M.functions()) {
+    if (F.getName().startswith(SYCL_GET_SCALAR_2020_SPEC_CONST_VAL) ||
+        F.getName().startswith(SYCL_GET_COMPOSITE_2020_SPEC_CONST_VAL))
+      return true;
+  }
+
+  return false;
+}
+
+/// Function generates a copy of Module and replaces all spec constant uses
+/// memcpy intrinsic.
+/// If the module doesn't contain spec constants then nullptr is returned.
+std::unique_ptr<Module>
+llvm::generateDeviceImageWithDefaultSpecConstants(const Module &M) {
+  if (!checkModuleContainsSpecConsts(M))
+    return nullptr;
+
+  std::unique_ptr<Module> M2 = CloneModule(M);
+  std::vector<Function *> SpecConstDeclarations;
+  std::vector<CallInst *> CIs;
+  for (Function &F : M2->functions()) {
+    if (!F.isDeclaration())
+      continue;
+
+    if (!F.getName().startswith(SYCL_GET_SCALAR_2020_SPEC_CONST_VAL) &&
+        !F.getName().startswith(SYCL_GET_COMPOSITE_2020_SPEC_CONST_VAL))
+      continue;
+
+    SpecConstDeclarations.push_back(&F);
+    for (User *U : F.users())
+      if (auto *CI = dyn_cast<CallInst>(U))
+        CIs.push_back(CI);
+  }
+
+  replaceSpecConstsWithDefaultValue(*M2, CIs);
```

Comment:
This kind of replicates existing pass logic. I would suggest that we replace "bool SetValAtRT = true" with some `enum`, which would have 3 possible values:
- `native` - emit SPIR-V intrinsics
- `emulation` - use kernel arguments
- `default` - replace with default values

Then, we will simply be able to run the pass in different modes. It should be updated to have a 3rd main branch, which we kind of had for SYCL 1.2.1 (in AOT mode, they values were replaced with defaults)